### PR TITLE
Prepare GoodDollarStaking for audit

### PR DESCRIPTION
**Pull request by Shloyem**

### CODE REVIEW

File: contracts/governance/GoodDollarStaking.sol [hunk 1]
```diff
@@ -244,17 +244,36 @@ contract GoodDollarStaking is
...
+    _goodReward = getUserPendingReward(
+			address(this),
+			0,
+			block.number,
+			_user
+		);
+    
+    // g$ rewards
+    _updateReward(_user);
+    _gdReward = stakersInfo[_user].reward;
```
Comment (Shloyem) [inline contracts/governance/GoodDollarStaking.sol:267]: Could be done just with earned() function, if we revert previous change in FixedAPY

stakersInfo[_account].reward **+**= uint128(earned(_account)); // added +
earned(){ .... calculation... ~~+ stakersInfo[_account].reward~~

but not a big deal

File: contracts/governance/GoodDollarStaking.sol [hunk 2]
```diff
@@ -234,27 +234,52 @@ contract GoodDollarStaking is
 		_setAPY(_interestRatePerBlock);
 	}
 
-	/// @dev helper function for multibase
-	function getRewardsPerBlock() public view returns (uint256) {
-		return rewardsPerBlock[address(this)];
+	/// @dev helper function for multibase and FixedAPYRewards
+	function getRewardsPerBlock() public view returns (uint256 _goodRewardPerBlock, int128 _gdRewardPerBlockX64) {
```
Comment (sirpy) [inline contracts/governance/GoodDollarStaking.sol:238]:
```suggestion
	function getRewardsPerBlock() public view returns (uint256 goodRewardPerBlock, uint256 gdInterestRatePerBlock) {
```
prefix `_` is only for method parameters
you should return the interestrate in regular uint256, that means that you need to multiply interestrateX64 by precision (`.mulu`)

File: contracts/governance/GoodDollarStaking.sol (hunk 1 above)
Comment (sirpy) [reply contracts/governance/GoodDollarStaking.sol:263]: you can not simply convert a "view" function into a non-view function
you need to calculate what is the user pending reward but not update it

File: contracts/governance/GoodDollarStaking.sol [hunk 3]
```diff
@@ -234,27 +234,52 @@ contract GoodDollarStaking is
...
+		public
+		view
+		returns (uint256 _goodRewardPerShare, uint128 _gdRewardPerShare)
+	{
+		_goodRewardPerShare = super.totalRewardsPerShare(address(this));
+		_gdRewardPerShare = rewardPerTokenStored;
 	}
 
-	/// @dev helper function for multibase
+  /// @dev helper function for multibase
```
Comment (sirpy) [inline contracts/governance/GoodDollarStaking.sol:276]: whats going on with the indentations? fix your editor to automatically indent and prettier on save

File: contracts/staking/utils/StakingRewardsFixedAPY.sol [hunk 4]
```diff
@@ -56,18 +56,19 @@ contract StakingRewardsFixedAPY {
...
 		//earned perToken = earnedInTimeSpan*PRECISION/totalStaked
 		//PRECISION cancels out
 		int128 compound = interestRatePerBlockX64.pow(
 			block.number - lastUpdateBlock
 		);
-		uint256 earnedInterest = compound.mulu(principle);
-		principle += earnedInterest;
+		uint256 principleWithInterest = compound.mulu(principle);
+		principle = principleWithInterest;
+		uint256 earnedInterest = principleWithInterest - principle;
```
Comment (sirpy) [inline contracts/staking/utils/StakingRewardsFixedAPY.sol:71]: thats going to be 0

**Review by sirpy — CHANGES_REQUESTED** [review]

File: contracts/governance/GoodDollarStaking.sol (hunk 1 above)
Comment (Shloyem) [reply contracts/governance/GoodDollarStaking.sol]: Understood, thx

File: contracts/governance/GoodDollarStaking.sol (hunk 3 above)
Comment (Shloyem) [reply contracts/governance/GoodDollarStaking.sol]: I don't know, sometimes it adds spaces although it doesn't look like that locally.
Added format on save and will fix it for good

File: scripts/multichain-deploy/gdSavings-deploy.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
+/***
+ * complete deploy of the new G$ savings concept
+ * steps required:
+ * sidechain
+ * 1. deploy GoodDollarMintBurnWrapper
+ * 2. deploy G$Savings
+ * 5. create proposal to:
```
Comment (sirpy) [inline scripts/multichain-deploy/gdSavings-deploy.ts:7]:
```suggestion
 * 3. create proposal to:
```

File: scripts/multichain-deploy/gdSavings-deploy.ts [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+import { deployWrapper } from "./multichainWrapper-deploy";
+import {
+  CompoundVotingMachine,
+  GoodDollarMintBurnWrapper,
+  Controller,
+  NameService
+} from "../../types";
+const { name: networkName } = network;
+
+const BLOCKS_PER_YEAR = (12 * 60 * 24 * 356).toString();
```
Comment (sirpy) [inline scripts/multichain-deploy/gdSavings-deploy.ts:36]:
```suggestion
const BLOCKS_PER_YEAR = (12 * 60 * 24 * 365).toString();
```

File: test/staking/StakingRewardsFixedAPY.test.ts [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+      .div(statsBeforeUndo.totalShares.add(newSharesOnUndo));
+    expect(areApproxEqual(statsAfterUndo.avgDonationRatio, expectedGlobalAvgRatio));
+  });
+
+  it("Should fail to undo exceeding reward", async () => {
+    const { staking } = await waffle.loadFixture(fixture_1year);
+    await staking.withdraw(staker3.address, 500);
+    const errorTx = await staking.undoReward(staker3.address, 501).
+      catch(e => e);
+    expect(errorTx).to.be.an("error");
```
Comment (sirpy) [inline test/staking/StakingRewardsFixedAPY.test.ts:470]:
```suggestion
    await expect(staking.undoReward(staker3.address, 501)).to.be.reverted
```

**Review by sirpy — APPROVED** [review]

**Pull request merged**